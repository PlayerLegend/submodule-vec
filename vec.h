#ifndef FLAT_INCLUDES
#define FLAT_INCLUDES
#endif

typedef int ivec;
typedef float fvec;

//#define FVEC_EPSILON 0.000001
#define FVEC_EPSILON 0.00001
