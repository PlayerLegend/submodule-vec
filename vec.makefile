C_PROGRAMS += test/vec
C_PROGRAMS += test/mat4

run-tests: run-vec-tests
run-vec-tests:
	DEPENDS=vec-tests sh run-tests.sh test/vec test/mat4

depend: vec-depend
vec-depend:
	sh makedepend.sh src/vec/vec.makefile

vec-tests: test/vec
vec-tests: test/mat4

test/vec: LDLIBS += -lm
test/vec: src/vec/test/vec.test.o

test/mat4: LDLIBS += -lm
test/mat4: src/vec/test/mat4.test.o
test/mat4: src/vec/mat4.o
test/mat4: src/log/log.o

tests: vec-tests
# DO NOT DELETE

src/vec/view.o: src/vec/vec.h src/vec/vec3.h src/vec/vec4.h src/vec/view.h
src/vec/view.o: src/keyargs/keyargs.h src/vec/mat4.h
src/vec/vec3.o: src/vec/vec.h src/vec/vec3.h
src/vec/test/mat4.test.o: src/vec/vec.h src/vec/vec3.h src/keyargs/keyargs.h
src/vec/test/mat4.test.o: src/vec/vec4.h src/vec/mat4.h src/log/log.h
src/vec/test/vec.test.o: src/vec/vec.h src/vec/vec2.h src/vec/vec3.h
src/vec/mat4.o: src/vec/vec.h src/vec/vec2.h src/vec/vec3.h src/vec/vec4.h
src/vec/mat4.o: src/keyargs/keyargs.h src/vec/mat4.h src/log/log.h
src/vec/vec4.o: src/vec/vec.h src/vec/vec3.h src/vec/vec4.h
