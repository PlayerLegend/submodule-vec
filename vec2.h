#ifndef FLAT_INCLUDES
#define FLAT_INCLUDES
#include "vec.h"
#endif

#define vec2(type) struct { type x; type y; }
#define vec2_add(a,b) { (a).x += (b).x; (a).y += (b).y; }
#define vec2_subtract(a,b) { (a).x -= (b).x; (a).y -= (b).y; }
#define vec2_scale(v,s) { (v).x *= s; (v).y *= s; }
#define vec2_subtract_init(a,b) { (a).x - (b).x, (a).y - (b).y }

typedef vec2(ivec) ivec2;
typedef vec2(fvec) fvec2;
