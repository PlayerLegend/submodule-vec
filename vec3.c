#include <math.h>
#define FLAT_INCLUDES
#include "vec.h"
#include "vec3.h"
#include "../log/log.h"

fvec fvec3_vlen (const fvec3 * a)
{
    return sqrt (a->x * a->x + a->y * a->y + a->z * a->z);
}

void fvec3_normalize (fvec3 * result, const fvec3 * input)
{
    float length = fvec3_vlen (input);

    if (length < FVEC_EPSILON)
    {
	*result = (fvec3){0};
    }
    else
    {
	float inv_length = 1.0 / length;

	*result = (fvec3){ vec3_scale_init (*input, inv_length) };
    }
}
