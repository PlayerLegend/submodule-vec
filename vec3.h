#ifndef FLAT_INCLUDES
#define FLAT_INCLUDES
#include "vec.h"
#endif

#define vec3(type) struct { type x; type y; type z; }
#define vec3_add(a,b) { (a).x += (b).x; (a).y += (b).y; (a).z += (b).z; }
#define vec3_subtract(a,b) { (a).x -= (b).x; (a).y -= (b).y; (a).z -= (b).z; }
#define vec3_subtract_init(a,b) { (a).x - (b).x, (a).y - (b).y, (a).z - (b).z }
#define vec3_add_init(a,b) { (a).x + (b).x, (a).y + (b).y, (a).z + (b).z }
#define vec3_scale(a,s) { (a).x *= s; (a).y *= s; (a).z *= s; }
#define vec3_dot(a,b) ( (a).x * (b).x + (a).y * (b).y + (a).z * (b).z )
#define vec3_cross(a,b) { (a).y * (b).z - (a).z * (b).y, (a).z * (b).x - (a).x * (b).z, (a).x * (b).y - (a).y * (b).x }
#define vec3_line_point_init(position,direction,scale) { (position).x + (direction).x * (scale), (position).y + (direction).y * (scale), (position).z + (direction).z * (scale) }
#define vec3_initializer(a) { (a).x, (a).y, (a).z }
#define vec3_scale_init(a,s) { (a).x * (s), (a).y * (s), (a).z * (s) }

#define VEC3_FORMSPEC "%f %f %f"
#define VEC3_FORMSPEC_ARG(v) (v).x, (v).y, (v).z

typedef vec3(ivec) ivec3;
typedef vec3(fvec) fvec3;

fvec fvec3_vlen (const fvec3 * v);
