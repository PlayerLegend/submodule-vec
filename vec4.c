#include <math.h>
#define FLAT_INCLUDES
#include "vec.h"
#include "vec3.h"
#include "vec4.h"
#include <assert.h>
#include "../log/log.h"

void fvec4_setup_rotation_quaternion (fvec4 * q, const fvec3 * axis)
{
    fvec length = (axis->x || axis->y || axis->z) ? fvec3_vlen (axis) : 1;
    
    fvec3 axis_normal =
	{
	    .x = axis->x / length,
	    .y = axis->y / length,
	    .z = axis->z / length
	};

    float cos_angle_2 = cos(length / 2);
    float sin_angle_2 = sin(length / 2);

    q->x = sin_angle_2 * axis_normal.x;
    q->y = sin_angle_2 * axis_normal.y;
    q->z = sin_angle_2 * axis_normal.z;

    q->w = cos_angle_2;
    //q->w = 1;
}

void fvec4_normalize (fvec4 * result, const fvec4 * input)
{
    const fvec length = sqrt(input->x * input->x + input->y * input->y + input->z * input->z + input->w * input->w);
    assert (length > FVEC_EPSILON);
    const fvec invlength = 1.0 / length;
    *result = (fvec4){ vec4_scale_init(*input, invlength) };
}

void fvec4_quaternion_multiply (fvec4 * result, const fvec4 * a, const fvec4 * b)
{
    result->w = a->w * b->w - a->x * b->x - a->y * b->y - a->z * b->z;
    result->x = a->w * b->x + a->x * b->w + a->y * b->z - a->z * b->y;
    result->y = a->w * b->y - a->x * b->z + a->y * b->w + a->z * b->x;
    result->z = a->w * b->z + a->x * b->y - a->y * b->x + a->z * b->w;
}

void fvec4_quaternion_angle_delta (fvec4 * result, const fvec3 * begin, const fvec3 * end)
{
    result->w = fvec3_vlen(begin) * fvec3_vlen(end) + vec3_dot(*begin, *end);

    assert (result->w > -FVEC_EPSILON);

    if (result->w < FVEC_EPSILON)
    {
	fvec3 perpendicular = { vec3_cross_init_unit_x(*begin) };
	fvec perpendicular_length = fvec3_vlen(&perpendicular);

	if (perpendicular_length < FVEC_EPSILON)
	{
	    perpendicular = (fvec3){ vec3_cross_init_unit_y(*begin) };
	    perpendicular_length = fvec3_vlen(&perpendicular);
	    assert (perpendicular_length >= FVEC_EPSILON);
	}

	fvec scale = 3.14159 / perpendicular_length;

	vec3_scale(perpendicular, scale);

	fvec4_setup_rotation_quaternion(result, &perpendicular);
    }
    else
    {
	result->alias_vec3 = (fvec3){ vec3_cross_init(*end, *begin) };
    }

    fvec4_normalize(result, result);
}

void fvec4_apply_rotation_quaternion (fvec4 * target, const fvec4 * apply)
{
    /*fvec4 apply_inverse =
	{
	    .w = target->w,
	    .x = -target->x,
	    .y = -target->y,
	    .z = -target->z,
	    };*/
    
    fvec4 tmp = *target;

    //vec4_quaternion_multiply(&tmp, apply, target);

    fvec4_quaternion_multiply (target, &tmp, apply);
    
    float scale = 1.0 / vec4_vlen(target);
	
    vec4_scale (*target, scale);
}

#define qx q->z
#define qy q->y
#define qz q->x
#define qw q->w
//float vec4_quaternion_pitch (const fvec4 * q)
float vec4_quaternion_yaw (const fvec4 * q)
{
    return atan2 (2 * (qy * qw - qx * qz), 1 - 2 * (qy * qy + qz * qz));
}

//float vec4_quaternion_yaw (const fvec4 * q)
float vec4_quaternion_pitch (const fvec4 * q)
{
    return asin (2 * (qx * qy + qz * qw));
}

//float vec4_quaternion_roll (const fvec4 * q)
float vec4_quaternion_roll (const fvec4 * q)
{
    return atan2 (2 * (qx * qw - qy * qz), 1 - 2 * (qx * qx + qz * qz));
}

float vec4_vlen (const fvec4 * v)
{
    return sqrt (v->x * v->x + v->y * v->y + v->z * v->z + v->w * v->w);
}

void fvec4_apply_rotation_axis (fvec4 * target, const fvec3 * axis)
{
    fvec4 new_quaternion;
    fvec4_setup_rotation_quaternion(&new_quaternion, axis);
    fvec4_apply_rotation_quaternion(target, &new_quaternion);
}

void fvec4_apply_rotation_axis_angle (fvec4 * target, const fvec3 * axis, fvec angle)
{
    fvec axis_length = fvec3_vlen (axis);

    if (axis_length == 0)
    {
	return;
    }

    fvec axis_scale = angle / axis_length;

    fvec3 axis_fix = { vec3_scale_init(*axis, axis_scale) };

    fvec4_apply_rotation_axis(target, &axis_fix);
}

fvec vec4_quaternion_roll_mod (const fvec4 * q)
{
    fvec roll = vec4_quaternion_roll (q);
    fvec pi = 3.14159;

    if (roll < -0.5)
    {
	return roll + pi;
    }
    else if (roll > 0.5)
    {
	return roll - pi;
    }
    else
    {
	return roll;
    }
}

void fvec4_hamiltonian_product (fvec4 * result, const fvec4 * a, const fvec4 * b)
{
    *result = (fvec4)
	{
	    .w = a->w * b->w - a->x * b->x - a->y * b->y - a->z * b->z,
	    .x = a->w * b->x + a->x * b->w + a->y * b->z - a->z * b->y,
	    .y = a->w * b->y - a->x * b->z + a->y * b->w + a->z * b->x,
	    .z = a->w * b->z + a->x * b->y - a->y * b->x + a->z * b->w,
        };
}

void fvec4_quaternion_rotate (fvec3 * output, const fvec4 * q_quaternion, const fvec3 * input)
{
    fvec4 v_target = { .alias_vec3 = *input };
    
    fvec4 c_compliment =
	{
	    .w = q_quaternion->w,
	    .x = -q_quaternion->x,
	    .y = -q_quaternion->y,
	    .z = -q_quaternion->z,
	};

    fvec4 qv_hamiltonian;
    fvec4_hamiltonian_product(&qv_hamiltonian, q_quaternion, &v_target);

    fvec4 qvc_hamiltonian;
    fvec4_hamiltonian_product(&qvc_hamiltonian, &qv_hamiltonian, &c_compliment);

    *output = qvc_hamiltonian.alias_vec3;
}

/*void fvec4_quaternion_unrotate (fvec3 * output, const fvec4 * q_quaternion, const fvec3 * input)
{
    fvec4 v_target = { .alias_vec3 = *input };
    
    fvec4 c_compliment =
	{
	    .w = q_quaternion->w,
	    .x = -q_quaternion->x,
	    .y = -q_quaternion->y,
	    .z = -q_quaternion->z,
	};

    fvec4 cv_hamiltonian;
    vec4_hamiltonian_product(&cv_hamiltonian, &c_compliment, &v_target);

    fvec4 cvq_hamiltonian;
    vec4_hamiltonian_product(&cvq_hamiltonian, &cv_hamiltonian, q_quaternion);

    *output = cvq_hamiltonian.alias_vec3;
}
*/
