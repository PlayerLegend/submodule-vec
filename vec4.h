#ifndef FLAT_INCLUDES
#define FLAT_INCLUDES
#include "vec.h"
#include "vec3.h"
#endif

#define vec4(type) union { struct { type x; type y; type z; type w; }; type index[4]; }
#define vec4_add(a,b) { (a).x += (b).x; (a).y += (b).y; (a).z += (b).z; (a).w += (b).w; }
#define vec4_subtract(a,b) { (a).x -= (b).x; (a).y -= (b).y; (a).z -= (b).z; (a).w -= (b).w; }
#define vec4_scale(a,s) { (a).x *= s; (a).y *= s; (a).z *= s; (a).w *= s; }

typedef vec4(ivec) ivec4;
typedef vec4(fvec) fvec4;

void vec4_setup_rotation_quaternion (fvec4 * q, const fvec3 * axis);
void vec4_apply_rotation_quaternion (fvec4 * target, const fvec4 * apply);

void vec4_apply_rotation_axis (fvec4 * target, const fvec3 * axis);
void vec4_apply_rotation_axis_angle (fvec4 * target, const fvec3 * axis, fvec angle);

float vec4_quaternion_yaw (const fvec4 * q);
float vec4_quaternion_pitch (const fvec4 * q);
float vec4_quaternion_roll (const fvec4 * q);
fvec vec4_quaternion_roll_mod (const fvec4 * q);

float vec4_vlen (const fvec4 * v);
