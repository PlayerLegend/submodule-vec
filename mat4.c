#include <string.h>
#include <math.h>
#define FLAT_INCLUDES
#include "vec.h"
#include "vec2.h"
#include "vec3.h"
#include "vec4.h"
#include "../keyargs/keyargs.h"
#include "mat4.h"
#include "../log/log.h"

/*void apply_translation (mat4 input, fvec3 * tl)
{
#define tl_matrix_unwrap(row) input[12 + row] = input[0 + row]*tl->x + input[4 + row]*tl->y + input[8 + row]*tl->z + input[12 + row];
    tl_matrix_unwrap(0);
    tl_matrix_unwrap(1);
    tl_matrix_unwrap(2);
    tl_matrix_unwrap(3);
    }*/

keyargs_define(mat4_setup_rotation_matrix)
{
    // https://en.wikipedia.org/wiki/Quaternions_and_spatial_rotation
     
    float s = 1 / (+ args.quaternion->x * args.quaternion->x
		   + args.quaternion->y * args.quaternion->y
		   + args.quaternion->z * args.quaternion->z
		   + args.quaternion->w * args.quaternion->w);

    s = 1;
    
    float s2 = 2 * s;

    *args.result = (mat4){
	.column = {
	    {
		.x  = 1 - s2 * (args.quaternion->y * args.quaternion->y + args.quaternion->z * args.quaternion->z),
		.y  = s2 * (args.quaternion->x * args.quaternion->y + args.quaternion->z * args.quaternion->w),
		.z  = s2 * (args.quaternion->x * args.quaternion->z - args.quaternion->y * args.quaternion->w),
		.w  = 0,
	    },
	    {
		.x  = s2 * (args.quaternion->x * args.quaternion->y - args.quaternion->z * args.quaternion->w),
		.y  = 1 - s2 * (args.quaternion->x * args.quaternion->x + args.quaternion->z * args.quaternion->z),
		.z  = s2 * (args.quaternion->y * args.quaternion->z + args.quaternion->x * args.quaternion->w),
		.w  = 0,
	    },
	    {
		.x  = s2 * (args.quaternion->x * args.quaternion->z + args.quaternion->y * args.quaternion->w),
		.y  = s2 * (args.quaternion->y * args.quaternion->z - args.quaternion->x * args.quaternion->w),
		.z = 1 - s2 * (args.quaternion->x * args.quaternion->x + args.quaternion->y * args.quaternion->y),
		.w = 0,
	    },
	    {
		.x = 0,
		.y = 0,
		.z = 0,
		.w = 1,
	    },
	},
    };
    
    

    /*
    mat4 a;
    a[0] = args.quaternion->w;
    a[1] = -args.quaternion->z;
    a[2] = args.quaternion->y;
    a[3] = -args.quaternion->x;

    a[4] = args.quaternion->z;
    a[5] = args.quaternion->w;
    a[6] = -args.quaternion->x;
    a[7] = -args.quaternion->y;

    a[8] = -args.quaternion->y;
    a[9] = args.quaternion->x;
    a[10] = args.quaternion->w;
    a[11] = -args.quaternion->z;

    a[12] = args.quaternion->x;
    a[13] = args.quaternion->y;
    a[14] = args.quaternion->z;
    a[15] = args.quaternion->w;

    mat4 b;
    b[0] = args.quaternion->w;
    b[1] = -args.quaternion->z;
    b[2] = args.quaternion->y;
    b[3] = args.quaternion->x;

    b[4] = args.quaternion->z;
    b[5] = args.quaternion->w;
    b[6] = -args.quaternion->x;
    b[7] = args.quaternion->y;

    b[8] = -args.quaternion->y;
    b[9] = args.quaternion->x;
    b[10] = args.quaternion->w;
    b[11] = args.quaternion->z;
    
    b[12] = -args.quaternion->x;
    b[13] = -args.quaternion->y;
    b[14] = -args.quaternion->z;
    b[15] = args.quaternion->w;

    mat4_multiply(args.result, &a, &b);*/

    /*(*args.result)[0] = 2 * (q.w * q.w + q.x * q.x) - 1;
    (*args.result)[1] = 2 * (q.w * q.y + q.w * q.z);
    (*args.result)[2] = 2 * (q.x * q.z - q.w * q.y);
    (*args.result)[3] = 0;
    
    (*args.result)[0] = 2 * (q.x * q.y - q.w * q.z);
    (*args.result)[0] = 2 * (q.w * q.w + q.y * q.y) - 1;
    (*args.result)[0] = 2 * (q.y * q.z + q.w * q.x);
    (*args.result)[0] = 0;
    
    (*args.result)[0] = 2 * (q.x * q.z + q.w * q.y);
    (*args.result)[0] = 2 * (q.y * q.z - q.w * q.x);
    (*args.result)[0] = 2 * (q.w * q.w + q.z * q.z) - 1;
    (*args.result)[0] = 0;
    
    (*args.result)[12] = 0;
    (*args.result)[13] = 0;
    (*args.result)[14] = 0;
    (*args.result)[15] = 1;*/
}

keyargs_define(mat4_setup_projection_matrix)
{
    /* https://www.khronos.org/registry/OpenGL-Refpages/gl2.1/xhtml/gluPerspective.xml */

    float f = args.fovy / 2;
    f = cos (f) / sin (f);

    *args.result = (mat4)
	{
	    .column =
	    {
		{
		    .x = f / args.aspect,
		},
		{
		    .y = f,
		},
		{
		    .z = (args.far + args.near) / (args.near - args.far),
		    .w = -1,
		},
		{
		    .z = (2 * args.far * args.near) / (args.near - args.far),
		},
	    }
	};
}

keyargs_define(mat4_setup_translation_matrix)
{
    *args.result = (mat4)
    {
	.column =
	{
	    { .x = 1 },
	    { .y = 1 },
	    { .z = 1 },
	    {
		.x = args.translation.x,
		.y = args.translation.y,
		.z = args.translation.z,
		.w = 1,
	    }
	}
    };
}

keyargs_define(mat4_setup_scale_matrix)
{
    *args.result = (mat4)
    {
	.column =
	{
	    { .x = args.scale.x },
	    { .y = args.scale.y },
	    { .z = args.scale.z },
	    { .w = 1 },
	},
    };
}

/*keyargs_define(mat4_setup_transform_matrix)
{
    mat4 a, b, c;

    mat4_setup_rotation_matrix(.result = &b, .axis = args.axis);
    mat4_setup_scale_matrix (.result = &c, .scale = args.scale);
    
    mat4_multiply (&a, &b, &c);

    mat4_setup_translation_matrix (.result = &c, .translation = args.translation);

    mat4_multiply (&b, &c, &a);

    mat4_setup_projection_matrix(.result = &a, .fovy = args.fovy, .aspect = 1, .near = 0.01, .far = 1);

    mat4_multiply (args.result, &a, &b);
    }*/

/*inline static float matrix_multiply_dot (int row_index, int col_index, const mat4 * a, const mat4 * b)
{
    return
	+ a->column[0].index[row_index] * b->column[col_index].x
	+ a->column[1].index[row_index] * b->column[col_index].y
	+ a->column[2].index[row_index] * b->column[col_index].z
	+ a->column[3].index[row_index] * b->column[col_index].w;
    
    const fvec * row = (*a) + row_index;
    const fvec * col = (*b) + 4 * col_index;

    
    return
	+ row[0] * col[0]
	+ row[4] * col[1]
	+ row[8] * col[2]
	+ row[12] * col[3];
}*/

#define mat4_dot_column_vec(a, member, v)			\
    + (a).column[0].member * (v).x				\
    + (a).column[1].member * (v).y				\
    + (a).column[2].member * (v).z				\
    + (a).column[3].member * (v).w

#define mat4_dot_column(a, v)		\
	.x = mat4_dot_column_vec(a, x, v),	\
	    .y = mat4_dot_column_vec(a, y, v),	\
	    .z = mat4_dot_column_vec(a, z, v),	\
	    .w = mat4_dot_column_vec(a, w, v)

void mat4_multiply_column_vec (fvec4 * result, const mat4 * a, const fvec4 * b)
{
    *result = (fvec4){ mat4_dot_column(*a, *b) };
}

void mat4_multiply (mat4 * result, const mat4 * a, const mat4 * b)
{
    *result = (mat4)
	{
	    .column =
	    {
		{ mat4_dot_column(*a, b->column[0]) },
		{ mat4_dot_column(*a, b->column[1]) },
		{ mat4_dot_column(*a, b->column[2]) },
		{ mat4_dot_column(*a, b->column[3]) },
	    }
	};
}

void mat4_transpose (mat4 * a)
{
    float tmp;

    float * xp;
    float * yp;
    
    for (int x = 0; x < 4; x++)
    {
	for (int y = 0; y < x; y++)
	{
	    xp = a->column[x].index + y;
	    yp = a->column[y].index + x;

	    //log_debug("1 %f %f", *xp, *yp);
	    tmp = *xp;
	    *xp = *yp;
	    *yp = tmp;
	    //log_debug("2 %f %f", *xp, *yp);
	}
    }
}

/*void mat4_multiply (mat4 * out, const mat4 * a, const mat4 * b)
  {
  int row, col;

  for (row = 0; row < 4; row += 1)
  {
  for (col = 0; col < 16; col += 4)
  {
  (*out)[col + row]
  = (*a)[row] * (*b)[col]
  + (*a)[row + 4] * (*b)[col + 1]
  + (*a)[row + 8] * (*b)[col + 2]
  + (*a)[row + 12] * (*b)[col + 3];
  }
  }
  }
*/

void mat4_setup_identity_matrix (mat4 * target)
{
    *target = (mat4)
    {
	.column =
	{
	    { .x = 1 },
	    { .y = 1 },
	    { .z = 1 },
	    { .w = 1 },
	}
    };
}
