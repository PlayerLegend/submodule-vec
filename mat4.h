#ifndef FLAT_INCLUDES
#define FLAT_INCLUDES
#include "vec.h"
#include "vec3.h"
#include "vec4.h"
#include "../keyargs/keyargs.h"
#endif

typedef fvec mat4[16];
typedef struct mat4_swap mat4_swap;
struct mat4_swap {
    int result_index;
    mat4 contents[3];
};

void mat4_swap_init_identity (mat4_swap * swap);
void add_translation_matrix (mat4 input, fvec3 tl);

void mat4_multiply (mat4 * result, const mat4 * a, const mat4 * b);

void mat4_multiply_right_vec (fvec4 * result, const mat4 * a, const fvec4 * b);

// based on the information found here https://www.khronos.org/registry/OpenGL-Refpages/gl2.1/xhtml/gluPerspective.xml
#define mat4_setup_projection_matrix(...) keyargs_call(mat4_setup_projection_matrix, __VA_ARGS__)
keyargs_declare(void, mat4_setup_projection_matrix,
		mat4 * result;
		fvec fovy;
		fvec aspect;
		fvec near;
		fvec far;);

#define mat4_setup_translation_matrix(...) keyargs_call(mat4_setup_translation_matrix, __VA_ARGS__)
keyargs_declare(void, mat4_setup_translation_matrix,
		mat4 * result;
		fvec3 translation;);

#define mat4_setup_scale_matrix(...) keyargs_call(mat4_setup_scale_matrix, __VA_ARGS__)
keyargs_declare(void, mat4_setup_scale_matrix,
		mat4 * result;
		fvec3 scale;);

#define mat4_setup_rotation_matrix(...) keyargs_call(mat4_setup_rotation_matrix, __VA_ARGS__)
keyargs_declare(void, mat4_setup_rotation_matrix,
		mat4 * result;
		const fvec4 * quaternion;);

#define mat4_setup_transform_matrix(...) keyargs_call(mat4_setup_transform_matrix, __VA_ARGS__)
keyargs_declare(void, mat4_setup_transform_matrix,
		mat4 * result;
		fvec3 translation;
		fvec3 scale;
		fvec3 axis;
		fvec fovy;);

void mat4_setup_identity_matrix (mat4 * a);

void mat4_transpose (mat4 * a);
