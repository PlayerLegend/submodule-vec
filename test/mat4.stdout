Testing transpose
Before transpose
5.000000 7.000000 9.000000 10.000000
2.000000 3.000000 3.000000 8.000000
8.000000 10.000000 2.000000 3.000000
3.000000 3.000000 4.000000 8.000000
Comparing matrix After transpose:
5.000000 2.000000 8.000000 3.000000
7.000000 3.000000 10.000000 3.000000
9.000000 3.000000 2.000000 4.000000
10.000000 8.000000 3.000000 8.000000
To matrix Expected matrix:
5.000000 2.000000 8.000000 3.000000
7.000000 3.000000 10.000000 3.000000
9.000000 3.000000 2.000000 4.000000
10.000000 8.000000 3.000000 8.000000
Matrices match!

Multiplying matrix:
5.000000 7.000000 9.000000 10.000000
2.000000 3.000000 3.000000 8.000000
8.000000 10.000000 2.000000 3.000000
3.000000 3.000000 4.000000 8.000000
With matrix:
3.000000 10.000000 12.000000 18.000000
12.000000 1.000000 4.000000 9.000000
9.000000 10.000000 12.000000 2.000000
3.000000 12.000000 4.000000 10.000000
Comparing matrix multiply result:
210.000000 267.000000 236.000000 271.000000
93.000000 149.000000 104.000000 149.000000
171.000000 146.000000 172.000000 268.000000
105.000000 169.000000 128.000000 169.000000
To matrix expectation:
210.000000 267.000000 236.000000 271.000000
93.000000 149.000000 104.000000 149.000000
171.000000 146.000000 172.000000 268.000000
105.000000 169.000000 128.000000 169.000000
Matrices match!

Multiplying matrix:
5.000000 7.000000 9.000000 10.000000
2.000000 3.000000 3.000000 8.000000
8.000000 10.000000 2.000000 3.000000
3.000000 3.000000 4.000000 8.000000
With column vector
2.000000 3.000000 5.000000 7.000000
Result:
2.000000 3.000000 5.000000 7.000000
